ksql-engine/src/main/java/io/confluent/ksql/util/AvroUtil.java

```diff
@@ -0,0 +1,174 @@
+import io.confluent.ksql.metastore.MetastoreUtil;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.AbstractStreamCreateStatement;
+import io.confluent.ksql.parser.tree.CreateStream;
+import io.confluent.ksql.parser.tree.CreateTable;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.parser.tree.TableElement;
+import io.confluent.ksql.serde.DataSource;
+
+public class AvroUtil {
+
+  private static final Logger log = LoggerFactory.getLogger(AvroUtil.class);
+
+  public static synchronized Pair<AbstractStreamCreateStatement, String> checkAndSetAvroSchema(AbstractStreamCreateStatement
+           abstractStreamCreateStatement,
+       Map<String, Object> streamsProperties, SchemaRegistryClient schemaRegistryClient) {
+    Map<String,Expression> ddlProperties = abstractStreamCreateStatement.getProperties();
+    if (!ddlProperties.containsKey(DdlConfig.VALUE_FORMAT_PROPERTY)) {
+      throw
+          new KsqlException(String.format("%s should be set in WITH clause of CREATE "
+                                          + "STRAEM/TABLE statement.", DdlConfig.VALUE_FORMAT_PROPERTY));
+    }
+    final String serde = StringUtil.cleanQuotes(
+        ddlProperties.get(DdlConfig.VALUE_FORMAT_PROPERTY).toString());
+    if (!serde.equalsIgnoreCase(DataSource.AVRO_SERDE_NAME)) {
+      return new Pair<>(abstractStreamCreateStatement, null);
+    }
+
+    String kafkaTopicName = StringUtil.cleanQuotes(
+        ddlProperties.get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY).toString());
+    try {
+      SchemaMetadata schemaMetadata;
+      if (abstractStreamCreateStatement.getProperties().containsKey(KsqlConstants.AVRO_SCHEMA_ID)) {
+        int schemaId = -1;
+        try {
+          schemaId = Integer.parseInt(StringUtil.cleanQuotes(abstractStreamCreateStatement.getProperties().get(KsqlConstants.AVRO_SCHEMA_ID).toString()));
+        } catch (NumberFormatException e) {
+          throw new KsqlException(String.format("Invalid schema id property: %s.",
+                                                abstractStreamCreateStatement.getProperties().get(KsqlConstants.AVRO_SCHEMA_ID).toString()));
+        }
+        schemaMetadata = schemaRegistryClient.getSchemaMetadata(
+            kafkaTopicName +KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX, schemaId);
+      } else {
+        schemaMetadata = schemaRegistryClient.getLatestSchemaMetadata(kafkaTopicName +
+                                                                      KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX);
+      }
+
+      String avroSchemaString = schemaMetadata.getSchema();
+      streamsProperties.put(DdlConfig.AVRO_SCHEMA, avroSchemaString);
+      Schema schema = SerDeUtil.getSchemaFromAvro(avroSchemaString);
+      if (abstractStreamCreateStatement.getElements().isEmpty()) {
+        abstractStreamCreateStatement = addAvroFields(abstractStreamCreateStatement, schema,
```

Comment:
we should avoid re-assigning to a variable. Makes the code much harder to reason about. Would be better if the param was final